packages/host/app/components/ai-assistant/formatted-message.gts

```diff
@@ -1,18 +1,98 @@
+import { fn } from '@ember/helper';
+import { on } from '@ember/modifier';
+import { service } from '@ember/service';
 import type { SafeString } from '@ember/template';
 import Component from '@glimmer/component';
 
+import { tracked } from '@glimmer/tracking';
+
+import { task } from 'ember-concurrency';
+import perform from 'ember-concurrency/helpers/perform';
+import { TrackedArray, TrackedObject } from 'tracked-built-ins';
+
+import ApplySearchReplaceBlockCommand from '@cardstack/host/commands/apply-search-replace-block';
 import CodeBlock from '@cardstack/host/modifiers/code-block';
 import { MonacoEditorOptions } from '@cardstack/host/modifiers/monaco';
 
+import type CardService from '@cardstack/host/services/card-service';
+import CommandService from '@cardstack/host/services/command-service';
+import LoaderService from '@cardstack/host/services/loader-service';
 import { type MonacoSDK } from '@cardstack/host/services/monaco-service';
 
+import ApplyButton from '../ai-assistant/apply-button';
+
 interface FormattedMessageSignature {
   sanitizedHtml: SafeString;
   monacoSDK: MonacoSDK;
   renderCodeBlocks: boolean;
 }
 
+interface CodeAction {
```

Comment:
Perhaps we should consider making this a full-fledged CodeAction class.

If we did and we made it container-aware, it could have all the service dependencies and the task defined on it and keep FormattedMessage as more of a UI component.

Reply:
Good idea, will try something like that. On the other hand I'm also working on [https://linear.app/cardstack/issue/CS-8124/stream-code-directly-into-monaco-code-editor-fix-flashing-during](https://linear.app/cardstack/issue/CS-8124/stream-code-directly-into-monaco-code-editor-fix-flashing-during) which might result in a refactor around this area so I'll keep this comment open until I figure that out